src/Dialect/ONNX/ONNXOps.cpp

```diff
@@ -4164,11 +4164,128 @@ LogicalResult ONNXHardSwishOp::inferShapes(
+//===----------------------------------------------------------------------===//
+
+namespace {
+bool areIfTypesMergeable(Type ifResultType, Type branchResultType) {
+  // ifResultType must be tensor/seq/opt type because that's checked in
+  // ONNXIfOp::verifyInvariantsImpl()
+  if (ShapedType ifShapedType = ifResultType.dyn_cast<ShapedType>()) {
+    if (ShapedType branchShapedType = branchResultType.dyn_cast<ShapedType>()) {
+      return ifShapedType.getElementType() == branchShapedType.getElementType();
+    } else {
+      return false;
+    }
+  }
+  if (SeqType ifSeqType = ifResultType.dyn_cast<SeqType>()) {
+    if (SeqType branchSeqType = branchResultType.dyn_cast<SeqType>()) {
+      return areIfTypesMergeable(
+          ifSeqType.getElementType(), branchSeqType.getElementType());
+    } else {
+      return false;
+    }
+  }
+  if (OptType ifOptType = ifResultType.dyn_cast<OptType>()) {
+    if (OptType branchOptType = branchResultType.dyn_cast<OptType>()) {
+      return areIfTypesMergeable(
+          ifOptType.getElementType(), branchOptType.getElementType());
+    } else {
+      return false;
+    }
+  }
+  llvm_unreachable("areIfTypesMergeable called with non tensor/seq/opt type");
+}
+
+// Pre-condition: mergeableIfTypes(ifTy, lhs) && mergeableIfTypes(ifTy, rhs)
+Type mergeIfTypes(Type lhs, Type rhs) {
+  // All asserts below are checked in mergeableIfTypes().
+  if (ShapedType lhsShapedType = lhs.dyn_cast<ShapedType>()) {
+    ShapedType rhsShapedType = rhs.cast<ShapedType>();
+    Type elementType = lhsShapedType.getElementType();
+    assert(elementType == rhsShapedType.getElementType() &&
+           "tensor element types mismatch");
+    if (lhsShapedType.hasRank() && rhsShapedType.hasRank() &&
+        lhsShapedType.getRank() == rhsShapedType.getRank()) {
+      int64_t rank = lhsShapedType.getRank();
+      auto lhsShape = lhsShapedType.getShape();
+      auto rhsShape = rhsShapedType.getShape();
+      SmallVector<int64_t, 4> shape;
+      for (int64_t i = 0; i < rank; ++i) {
+        shape.push_back(lhsShape[i] == rhsShape[i] ? lhsShape[i] : -1);
+      }
+      return RankedTensorType::get(shape, elementType);
+    } else {
+      return UnrankedTensorType::get(elementType);
+    }
+  }
+  if (SeqType lhsSeqType = lhs.dyn_cast<SeqType>()) {
+    SeqType rhsSeqType = rhs.cast<SeqType>();
+    int64_t length = lhsSeqType.getLength() == rhsSeqType.getLength()
+                         ? lhsSeqType.getLength()
+                         : -1;
+    return SeqType::get(
+        mergeIfTypes(lhsSeqType.getElementType(), rhsSeqType.getElementType()),
+        length);
+  }
+  if (OptType lhsOptType = lhs.dyn_cast<OptType>()) {
+    OptType rhsOptType = rhs.cast<OptType>();
+    return OptType::get(
+        mergeIfTypes(lhsOptType.getElementType(), rhsOptType.getElementType()));
+  }
+  llvm_unreachable("mergeIfTypes called with non tensor/seq/opt type");
+}
+} // namespace
+
+LogicalResult ONNXIfOp::verify() {
```

Comment:
I am wondering whether the type of if result should be a super set of type of then branch and else branch. Or in another way, are we determine 1) T1 and T2 are mergable or 2) T1 contains T2?

Reply:
I don't think super set would match the the ONNX documentation for the If op which says "The return values in the `then_branch` and `else_branch` must be of the same data type".